demo/simulation-data-demo.R

```diff
@@ -0,0 +1,220 @@
+  Getw3 <- function (w2){
+    if (w2 == 1) {
+       min_val = 1.7
+       max_val = 2.4
+     }
+  else {
+       min_val = 1.4
+       max_val = 2.0
+     }
+  w3<-runif(1,min = min_val,max = max_val)
+  return(w3)
+  }  
+  
+  wv<-Vectorize(Getw3)
+  w3<-wv(w2)
+  
+  
+  A<-rbinom(n, size=1, prob=plogis(-0.4+0.1*w2+0.15*w3+0.15*w2*w3))
+  
+  distribution_x<-function(x,max_val, minx, maxx){round((x-minx)/(maxx-minx)*max_val)}
+  
+  noise<-rnorm(n,mean=0, sd=0.1)
+ 
+  Y <- (-1+-A-0.1*w1+0.4*w2+0.3*w3)+noise
+  #counterfactual
+  YA0 <-(-1-0-0.1*w1+0.4*w2+0.3*w3)+noise
+  YA1 <-(-1-1-0.1*w1+0.4*w2+0.3*w3)+noise
+  
+  minx<- min(c(Y, YA0, YA1))
+  maxx<- max(c(Y, YA0,YA1))
+  #browser()
+  Y <-distribution_x(Y,100,minx,maxx)
+  YA0 <-distribution_x(YA0,100,minx,maxx)
+  YA1 <-distribution_x(YA1,100,minx,maxx)
+  
+  data<-data.frame(Block,w1,w2,w3,A,Y,YA0,YA1,minx,maxx)
+  data$Patient_id <- seq.int(nrow(data))
+  return (data)
+  
+  
+}
+
+generateLagData<-function (simData_t0,ptn_id,from_block,to_block,prob_w2,n) {
+    for(j in from_block:to_block) {
+      if (j == 1){
+        
+         row_dag_df_0<-data.frame(filter(simData_t0,Patient_id==ptn_id))
+         row_dag<-row_dag_df_0
+      } else {row_dag <- row_dag_df_t}
+      
+      row_dag$Block<-row_dag$Block+1
+      row_dag$w1<-row_dag$w1
+      row_dag$Y<-row_dag$Y
+      Y_w2<-row_dag$Y
+      #browser()
+      ##calculate w2 depending on Y and previous W2
+      ##if Y> 50 increase if Y<50 decrease probability
+      ##if w2 = 0 decrease if w2= 1 increase probability
+      if (Y_w2<50){
+       delta_prob_w2<--(-(Y_w2/250))
+      } else {delta_prob_w2<-(Y_w2/500)}
+      
+      if (row_dag$w2==0){delta_prob_w2=delta_prob_w2-0.10
+      } else{delta_prob_w2=delta_prob_w2+0.10}
+      
+      row_dag$w2<-rbinom(n,size=1,prob=prob_w2+delta_prob_w2)
+      ##calculate w3 depending on w2
+      ## categorize using the PAL scale
+      ## function is neat, vectorization as in the first function is only needed when vectors are used
+      Getw3 <- function (w2){
+        if (w2 == 1) {
+          min_val = 1.7
+          max_val = 2.4
+        }
+        else {
+          min_val = 1.4
+          max_val = 2.0
+        }
+        w3<-runif(1,min = min_val,max = max_val)
+        return(w3)
+      }  
+      
+      row_dag$w3<-Getw3(row_dag$w2)
+      
+      #The use of A
+      noise<-rnorm(n,mean=0, sd=0.1)
+      
+      if (Y_w2<50){
+          A<-rbinom(n, size=1, prob=plogis(-0.4+0.1*row_dag$w2+0.15*row_dag$w3+0.15*row_dag$w2*row_dag$w3+noise))
+      } else { A<-rbinom(n, size=1, prob=plogis(-0.4+0.1*row_dag$w2+0.15*row_dag$w3+0.15*row_dag$w2*row_dag$w3+noise))}
+     
+      #n=1 when single patient_id is used
+      noise<-rnorm(n,mean=0, sd=0.1)
+      
+      row_dag$Y <- (-1-row_dag$A-0.1*row_dag$w1+0.4*row_dag$w2+0.3*row_dag$w3+Y_w2/100)+noise
+      #counter factual
+      row_dag$YA0 <- (-1-0-0.1*row_dag$w1+0.4*row_dag$w2+0.3*row_dag$w3+Y_w2/100)+noise
```

Comment:
volgens mij (maar verbeter me als het niet zo is) is die counter factual niet helemaal correct. Aangezien een vorige A effect heeft op de huidige W, en die weer effect heeft op de huidige A is het dus zo dat de vorige A niet perse constant hoeft te blijven (maar ik weet niet of dit een probleem hoeft te zijn)